Use hfmt to format Haskell code

wing328 opened the pull request:
### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.3.x`, `4.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

If the environment variable HFMT (e.g. $HOME/.local/bin/hfmt) is defined, use it to format the Haskell code.  

TODO: run hfmt to format Haskell servant code

jonschoning commented:
btw, are you generating the sample with `bin/haskell-http-client-petstore.sh` or `bin/openapi3/haskell-http-client-petstore.sh` ?

jonschoning commented:
not sure if i like having hfmt alter the output of the templates, but i'll look it over and see if it's more or less readable than the current output, a least for the haskell-http-client generator

wing328 commented:
@jonschoning thanks for the feedback. The formatting done by hfmt can be switched off by removing the environment variable "HFMT_PATH".

I agree with you that some users may not like the output of hfmt so the formatting should remain optional.

wing328 commented:
UPDATE: added `hfmt` to format Haskell Servant code, minor bug fix ("object" comparision) and enhancement (add "binary" type mapping) via https://github.com/OpenAPITools/openapi-generator/pull/968/commits/5d1dbde90abde3dcfc7b2a049da7f47913584423

wing328 commented:
@jonschoning thanks for reviewing the output. What about using `stylish-haskell` instead, which allows users to provide a config and seems to be more popular [2]?

Ref:
[1] https://github.com/jaspervdj/stylish-haskell#configuration
[2] https://www.reddit.com/r/haskell/comments/7bhsu0/what_code_formatter_are_you_using/

wing328 commented:
UPDATE: I've renamed the environment variable to "HASKELL_POST_PROCESS_FILE" to make it consistent across all generators. The user can define it with code formatter of their choice or using other tools to post process the auto-generated Haskell files.

jonschoning commented:
stylish-haskell has better formatting, but it appears to not handle all cases and generates errors: 

```
[main] ERROR o.o.c.l.HaskellHttpClientCodegen - Error running the command (stylish-haskell /home/jon/fs/git/openapi-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Client.hs). Exit value: 1
```

I like the approach of letting the user choose the code formatter as they wish.  
However, at least for haskell-http-client, I already put effort into making the templates generate well-formatted code.  How about we let the users choose to use HASKELL_POST_PROCESS_FILE if they wish, but we let the /samples remain the unadjusted output from the templates?

wing328 commented:
> [main] ERROR o.o.c.l.HaskellHttpClientCodegen - Error running the command (stylish-haskell -i /home/jon/fs/git/openapi-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Client.hs). Exit value: 1

I couldn't repeat the issue locally. Here are partial results:
```
[main] INFO  o.o.c.l.HaskellHttpClientCodegen - Successfully executed: /Users/williamcheng/.local/bin/stylish-haskell -i /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore.hs
[main] INFO  o.o.codegen.AbstractGenerator - writing file /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Client.hs
[main] INFO  o.o.c.l.HaskellHttpClientCodegen - Successfully executed: /Users/williamcheng/.local/bin/stylish-haskell -i /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Client.hs
[main] INFO  o.o.codegen.AbstractGenerator - writing file /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/API.hs
[main] INFO  o.o.c.l.HaskellHttpClientCodegen - Successfully executed: /Users/williamcheng/.local/bin/stylish-haskell -i /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/API.hs
[main] INFO  o.o.codegen.AbstractGenerator - writing file /Users/williamcheng/Code/graphql-generator/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Core.hs
```

Here is how I set the env variable:
```
export HASKELL_POST_PROCESS_FILE="/Users/williamcheng/.local/bin/stylish-haskell -i"
```

> How about we let the users choose to use HASKELL_POST_PROCESS_FILE if they wish, but we let the /samples remain the unadjusted output from the templates?

Definitely. I updated the samples to show you the difference made by the code formatter. The samples should be not be formatted automatically. I've updated this PR to restore the Haskell samples without using a code formatter.

jonschoning commented:
I like that.

I saw you used `./bin/haskell-http-client-petstore.sh` to generate the samples, where previously the samples were generated with `./bin/openapi3/haskell-http-client-petstore.sh`.  

which one would you prefer we generate the samples with? oas2 or oas3?

wing328 commented:
I still use oas2 version to update samples.

We may need to move some samples to another repository instead as the size of the repo will become too big if we include samples generated with OAS3 spec

jonschoning commented:
ok, i approve this then.. I will have to make some updates to the example app in the samples because it is compiling against the oas3 version.
